intermediate_source/named_tensor_tutorial.py

```diff
@@ -0,0 +1,518 @@
+# cases include flattening batch dimensions to send tensors into operators that
+# must take inputs with a certain number of dimensions (i.e., conv2d takes 4D input).
+#
+# To make these operation more semantically meaningful than view or reshape, we
+# introduce a new ``tensor.unflatten(dim, namedshape)`` method and update
+# ``flatten`` to work with names: ``tensor.flatten(dims, new_dim)``.
+#
+# ``flatten`` can only flatten adjacent dimensions but also works on
+# non-contiguous dims. One must pass into ``unflatten`` a **named shape**, which
+# is a list of ``(dim, size)`` tuples, to specify how to unflatten the dim. It
+# is possible to save the sizes during a ``flatten`` for ``unflatten`` but we
+# do not yet do that.
+
+imgs = imgs.flatten(['C', 'H', 'W'], 'features')
+print(imgs.names)
+
+imgs = imgs.unflatten('features', (('C', 2), ('H', 2), ('W', 2)))
+print(imgs.names)
+
+######################################################################
+# Autograd support
+# ----------------
+#
+# Autograd currently supports named tensors in a limited manner: autograd
+# ignores names on all tensors. Gradient computation is still correct but we
+# lose the safety that names give us. It is on the roadmap to introduce handling
+# of names to autograd.
+
+x = torch.randn(3, names=('D',))
+weight = torch.randn(3, names=('D',), requires_grad=True)
+loss = (x - weight).abs()
+grad_loss = torch.randn(3)
+loss.backward(grad_loss)
+
+print(weight.grad)  # Unnamed for now. Will be named in the future
+
+weight.grad.zero_()
+grad_loss = grad_loss.refine_names('C')
+loss = (x - weight).abs()
+# Ideally we'd check that the names of loss and grad_loss match but we don't yet.
+loss.backward(grad_loss)
+
+print(weight.grad)  # still unnamed
+
+######################################################################
+# Other supported (and unsupported) features
+# ------------------------------------------
+#
+# See here (link to be included) for a detailed breakdown of what is
+# supported with the 1.3 release, what is on the roadmap to be supported soon,
+# and what will be supported in the future but not soon.
+#
+# In particular, we want to call out three important features that are not
+# currently supported:
+#
+# - Retaining names when serializing or loading a serialized ``Tensor`` via
+#   ``torch.save``
+# - Multi-processing via ``torch.multiprocessing``
+# - JIT support; for example, the following will error
+
+imgs_named = torch.randn(1, 2, 2, 3 , names=('N', 'C', 'H', 'W'))
+
+@torch.jit.script
+def fn(x):
+    return x
+
+catch_error(lambda: fn(imgs_named))
+
+######################################################################
+# As a workaround, please drop names via `tensor = tensor.rename(None)`
+# before using anything that does not yet support named tensors.
+#
+# Longer example: Multi-headed attention
+# --------------------------------------
+#
+# Now we'll go through a complete example of implementing a common advanced

```

Comment:
Is it advanced?

Reply:
Perhaps, but that word is unnecessary :)